[Cloud7050] iP

Cloud7050 opened the pull request:
### Contrived Description

This PR is for the purpose of **receiving peer reviews** and demonstrating knowledge of [**GFMD**](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax). The latter has numerous requirements imposed, therefore, you may find this description to be:
- Highly verbose
- Oddly formatted
- Excessively long-winded

It's a bit like those stereotypical Bollywood films where the editor asks the director how many free transitions they want the movie to have, and the director replies `yes`.

1. ![](https://64.media.tumblr.com/4df14c045cd75909955586cc60c069e7/a8bd024b460ef1f3-d5/s400x600/7dbe0f84b1c792926d7513cfdd472ac0ede09f69.gif)
1. I can't find the main one I wanted to embed, but iirc it has people staring at each other out their car windows with a ton of cuts/transitions/effects applied back to back.

So... we use every syntax in the book. Kinda almost.

Now. Let's change topics so fast that you get literary whiplash. Here is how to ping your router on a loop when it starts acting up:

```bat
ping 192.168.0.1 -t -l 65500
```

And finally, allow me to end with a totally unrelated quote from 📖 "The Art of Computer Game Design" by Chris Crawford:

> Actions that are simple and obvious with other technologies become arcane with the computer. If I give you a bat and tell you that your goal in baseball is to hit the ball, you will have few problems deciding that you should swing the bat at the ball. A computer baseball game is not so easy to figure out. Do you press H for "hit" or S for "swing" or B for "bat"? Do you press the START key or press the joystick trigger? _Perhaps you should swing the joystick by its cable at the ball displayed on the television screen._

---

- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [x] a task list
- [x] an emoji
- [x] a blockquote
- [x] a hyperlink
- [x] inline code
- [x] some text formatting: bold, italic, strikethrough etc.

--- 

I could've just copy-pasted the template, I don't know why I spent so long making this. On the off chance anyone happens by this unremarkable PR... Hope you had a chuckle.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Cloud7050 committed:
chore(gitignore): Trim gitignore

Cloud7050 committed:
chore(md): Delete CONTRIBUTORS.md

Cloud7050 committed:
feat(): Rename Duke to Cloud

Cloud7050 committed:
feat(Cloud): Change default text to greeting/farewell

Cloud7050 committed:
feat(Cloud): Command handler, docs, bye command, echo unknown commands

Cloud7050 committed:
feat(Cloud): Commands to add TODOs, list TODOs

Cloud7050 committed:
feat(): Trim private docs, create Todo class

Cloud7050 committed:
feat(Cloud): Extract first word, handle empty, display isDone

Cloud7050 committed:
ref(Todo): Rename done to complete

Cloud7050 committed:
feat(Cloud): Mark/unmark with error handling, trim comments

Cloud7050 committed:
feat(Cloud): Simplify error checking, change todo display

Cloud7050 committed:
feat(todo): Move to todo package, simplify, create Deadline/Event

Cloud7050 committed:
style(): Fix accidental tabs, update VSCode settings for repo

Cloud7050 committed:
devfeat(Token): Create initial token handler

Cloud7050 committed:
ref(): Handle commands mostly as tokens

Cloud7050 committed:
ref(): TokenManager, trim setters

Cloud7050 committed:
devfeat(): Add type strings, Todo toString

Cloud7050 committed:
fix(TokenManager): Fix infinite removeTokens()

Cloud7050 committed:
ref(TokenManager): Store sub input without flag

Cloud7050 committed:
feat(todo): Different Todo string representations

Cloud7050 committed:
feat(Cloud): Create different Todos based on flags!

Cloud7050 committed:
chore(tests): Remove shell test script

Cloud7050 committed:
feat(src): Handle EOF, move packages

Also edit invalid mark error text

Cloud7050 committed:
chore(gitignore): Trim, edit for tests

Cloud7050 committed:
devfeat(tests): Update for semi-auto test

Cloud7050 committed:
ref(todo): Improve toString()

Reuse from Todo parent class via getBasicString()

Cloud7050 committed:
feat(src): Require command to add TODOs

Also:
• Reject unknown commands
• Add command aliases
• Add TokenManager#getDescription()
• Comment assumption and checks expected before using methods
• Rename Token#asInt() → toInt()

Cloud7050 committed:
devfeat(tests): Update inputs/outputs

Cloud7050 committed:
fix(): Use new MissingInputException, fix empty split() corner case

Cloud7050 committed:
ref(src): Use new @Nullable

Cloud7050 committed:
feat(): TokenManager child classes, empty sub input exception

Also enhances tests

Cloud7050 committed:
feat(): Intelligent space handling!

Performance was not much of a concern.

This is the kind of thing I'd have wanted for my Discord bots

Cloud7050 committed:
feat(): Add deletion, fix boundary error

Cloud7050 committed:
ref(src): Add CommandType enum!

AndrewJanong commented on src/com/cloud/chatbot/token/CommandManager.java:
```diff
@@ -0,0 +1,90 @@
+public class CommandManager extends TokenManager {
+    // Index by flag text instead of Token for faster retrieval
+    private HashMap<String, FlagManager> flagSets = new HashMap<>();
+
```
Shouldn't there be a comment header here before the constructor?

Cloud7050 replied:
Ah right, I'd forgotten that after the comment above the class itself. The constructor takes in a param which should be documented. I'll add that in subsequent commits.

AndrewJanong commented on src/com/cloud/chatbot/token/TokenManager.java:
```diff
@@ -0,0 +1,120 @@
+
+    public TokenManager() {}
+
+    protected static List<Token> stringToTokens(String s) {
```
Perhaps could use a verb for the function name? Something like `getTokensFromString`

Cloud7050 replied:
I get where you're coming from, though this is meant to be more of a utility function, thus it being a static method. I think names of the form "x to y" also see use elsewhere.

AndrewJanong commented on src/com/cloud/chatbot/Cloud.java:
```diff
@@ -0,0 +1,196 @@
+                )
+            );
+            break;
+        }
```
Good job on following the coding standard for switch statements 👍

AndrewJanong commented on src/com/cloud/chatbot/Cloud.java:
```diff
@@ -0,0 +1,196 @@
+        return new Todo(description);
+    }
+
+    private static @Nullable Integer verifyNumber(CommandManager manager) {
```
This function name does not suit with the return type. `verifyNumber` sounds like it would return a `boolean`.

Cloud7050 replied:
I see what you mean. This one is also more of a utility function. I'm a bit apprehensive about starting static methods with `get` since it sounds to me like an instance getter.

I named it so because part of its logic involves giving error feedback to the user, but at the end of the call, if the number is valid, then the callee needs to actually use said number (while also knowing whether or not it was valid), thus `@Nullable Integer` instead of just `int`.

I am open to better name suggestions.

Cloud7050 replied:
Come to think of it, if the method were meant to return a boolean, I think it would be called something like `isNumberValid()`. That thought occurred to me when reading this part of the code quality guide:

![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/44526554/cde4935d-2609-45ab-b76e-7fb5460de3bd)

AndrewJanong reviewed (commented):
## Overall Review
Overall, good job on following all code quality guidelines for the most part. I like how you keep things simple but still clear and readable. Do take a look at some of the things I pointed out. Keep it up 👍

siqirua commented on src/com/cloud/chatbot/Cloud.java:
```diff
@@ -0,0 +1,196 @@
+ * The chatbot's main class.
+ */
+public final class Cloud {
+    private static final Scanner SCANNER = new Scanner(System.in);
```
I like how you capitalize the static variable names

siqirua commented on src/com/cloud/chatbot/Cloud.java:
```diff
@@ -0,0 +1,196 @@
+            return null;
+        }
+
+        Token numberToken = manager.getToken(1);
+        if (!numberToken.isInt()) {
+            Cloud.say(
+                String.format(
+                    "\"%s\" is not a valid number.",
+                    numberToken.toString()
+                )
+            );
+            return null;
+        }
+
+        if (!numberToken.isValidNumber(Cloud.TODOS)) {
+            Cloud.say(
+                String.format(
+                    "TODO #%d does not exist.",
+                    numberToken.toInt()
+                )
+            );
+            return null;
+        }
```
I like how you break the line properly to improve readability!

siqirua commented on src/com/cloud/chatbot/token/CommandManager.java:
```diff
@@ -0,0 +1,90 @@
+
+            if (token.isFlag()) {
+                int endIndex = this.findFlagSetEnd(i);
+                List<Token> flagSet = this.removeTokens(i, endIndex);
```
```suggestion
                List<Token> flags = this.removeTokens(i, endIndex);
```
Is it better to name this variable as flags or flagList instead of flagSet since one might be confusing about the type of this variable (Set vs List)?

Cloud7050 replied:
I see what you mean. I hadn't considered the potential mixup with actual `Set`s. It's actually more so referring to this definition:

https://github.com/nus-cs2103-AY2324S1/ip/blob/abb83ad37dc5a8c4d916f4b414881e9e3ed8b10d/src/com/cloud/chatbot/token/CommandManager.java#L16-L18

Since I differentiate between flags and flag sets, the `List<Token>` in question is not really a flag or flags. Instead I prefer to call it a flag set (which includes 1 flag), and is related to the method call above of `.findFlagSetEnd()`.

siqirua commented on src/com/cloud/chatbot/token/TokenManager.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    protected static String tokensToString(List<Token> tokens) {
+        String joined = "";
```
Maybe can use a more meaningful name? It is a bit hard to infer what it is about simply by the variable name.

Cloud7050 replied:
The string starts empty but gets concatenated onto in subsequent lines, eventually being returned when done. In essence, it is a changing or accumulating temporary variable. If the logic gets more complex, I will consider changing the variable name or adding comments, maybe something like `joinedOutput` or documenting the protected method.

siqirua commented on src/com/cloud/chatbot/todo/Event.java:
```diff
@@ -0,0 +1,35 @@
+ * Represents an event TODO, which has both a starting and ending timestamp.
+ */
+public class Event extends Deadline {
+    private String timestampStart;
```
```suggestion
    private String startingTimestamp;
```
Would it make more sense to name the variable as startingTimestamp so it sounds more like a noun?

Cloud7050 replied:
Although I note that those aren't constants, I was actually going for this rule:

![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/44526554/8e9c8c14-816d-4137-b7ef-0fbf8d9fa0e7)

Ie `timestampStart` matching `timestampEnd`. I've experimented with similar naming schemes elsewhere. I guess I still haven't fully decided on which I prefer more.

I will keep this in mind, though I think once I update the type away from `String`, the new type will make it clear by its nature that those variables are meant to represent a timestamp, and I can just name the variables something like `start` and `end`.

siqirua reviewed (commented):
Overall, the code is easy to read and understand and I like how you follow the coding standards, especially the clear line break and method name. However, though the review should be focusing on the coding standard (so you may ignore this part if other aspects of the code is subject to change), maybe you can consider make a general Task interface and let Todo, Deadline and Event to be three parallel child class? For me, it does not really make sense that a event is also a deadline.

Overall good job! Keep it up!

Cloud7050 commented:
> maybe you can consider make a general Task interface and let Todo, Deadline and Event to be three parallel child class? For me, it does not really make sense that a event is also a deadline.

I suppose there are two ways to look at the inheritance.

There's the purely functional view, which I noticed and jumped on before actually scrolling down and seeing the `A-Inheritance` collapse further below the instructions (that talked about `Task` and inheritance). So by the time I read that part, I'd already implemented it "my way" and figured it was ok to stick with. From a properties point of view, `Todo` has no timestamps, `Deadline` has an end, and `Event` has both a start and an end. Due to the properties being a subset of one another, I went for not duplicating them.

Then there's the more real world interpretation, which I totally get. Yes, it does not really make sense for an event to strictly *be* a deadline. Though the way I ended up taking it was that they both have an ending timestamp.

To be honest, I'm not sure which is better. I do like how `Task` is an interface. Maybe it would even fit to have another subinterface like `EndableTask`. So as the project requirements change, I will keep this in mind and take it into account when touching these parts of the code down the line.

@siqirua

Cloud7050 committed:
ref(src): Remove redundant default constructors

Cloud7050 committed:
docs(src): Add missing public constructor docs

Cloud7050 committed:
chore(src): Restore src/main/java/ structure

Cloud7050 committed:
ref(src): TodoManager to prepare for file storage

Cloud7050 committed:
ref(): Rename TODO to Item, add Item parent class

Editor automatically recognises TODO as a dev comment.
This change disambiguates the two and prevents annoying highlighting.

Cloud7050 committed:
fix(test): Fix new src path in test's javac

Cloud7050 committed:
ref(src): Rename task to item in package

Cloud7050 committed:
chore(lib): Add org.json

Cloud7050 committed:
devfeat(src): Add export() item to JSON

Cloud7050 committed:
feat(src): Save to file with every item change

Cloud7050 committed:
feat(): Create directory if not exists. Gitignore

Cloud7050 committed:
docs(FileManager): JavaDoc

Cloud7050 committed:
feat(src): Read from file if exists

Cloud7050 committed:
fix(test): Update for lib & data folder

Cloud7050 committed:
ref(src): Enums for ItemType, JSON Key

Cloud7050 committed:
Merge branch 'branch-Level-7'

Cloud7050 committed:
FEAT(src): Handle real timestamps

Cloud7050 committed:
feat(tests): Update tests for new timestamps

Cloud7050 committed:
Merge branch 'branch-Level-8'

Cloud7050 committed:
ref(src): Extract Ui class

Cloud7050 committed:
ref(src): Extract Handler class

Cloud7050 committed:
ref(src): Singular package names

Cloud7050 committed:
ref(src): Split into Commands

Cloud7050 committed:
Merge branch 'add-gradle-support' of https://github.com/Cloud7050/ip

Cloud7050 committed:
style(test): Cosmetic capitalisation

Cloud7050 committed:
chore(gradle): Remove unix script

Cloud7050 committed:
chore(gitignore): Ignore .gradle/

Cloud7050 committed:
devfeat(): Set up working gradle

Cloud7050 committed:
devfeat(test): Set up runnable dummy tests

Cloud7050 committed:
devfeat(test): Add real tests

Cloud7050 committed:
chore(gradle): Tweaks for jar

Cloud7050 committed:
docs(src): Docs tweaks in root src dir

Cloud7050 committed:
docs(FileStorage): Tweak docs

Cloud7050 committed:
feat(src): FindCommand

Cloud7050 committed:
devfeat(test): Update & fix overall test script

Cloud7050 committed:
feat(FindCommand): No match message contains query

Cloud7050 committed:
Merge branch 'branch-A-JavaDoc'

Cloud7050 committed:
Merge branch 'branch-A-CodingStandard'

Cloud7050 committed:
Merge branch 'branch-Level-9'

Cloud7050 committed:
devfeat(): Add initial checkstyle

Cloud7050 committed:
chore(checkstyle): Tweak config

Cloud7050 committed:
chore(gitignore): Ignore bin/ from VSCode build

Cloud7050 committed:
style(src): Fix checkstyle issues

Cloud7050 committed:
Merge branch 'branch-A-CheckStyle'

Cloud7050 committed:
chore(gradle): Add JavaFX

Cloud7050 committed:
feat(src): Launch UI with label

Cloud7050 committed:
feat(Ui): Add initial nodes with neat styling

Cloud7050 committed:
deat(Ui): Initial event handlers

Cloud7050 committed:
chore(checkstyle): Allow lambda on nested newline

Cloud7050 committed:
feat(): Directional message rows with images

Cloud7050 committed:
feat(src): Add padding

Cloud7050 committed:
style(checkstyle): Tweak formatting

Cloud7050 committed:
ref(src): Rename to CloudApp

Cloud7050 committed:
feat(view): Add template fxml with slight edits

Cloud7050 committed:
feat(fxml): Create new components

Cloud7050 committed:
feat(fxml): Update views

Cloud7050 committed:
feat(): Run using FXML

Cloud7050 committed:
fix(fxml): Fix events. Set lower API version

Cloud7050 committed:
fix(fxml): Fix alignment width bug

Cloud7050 committed:
feat(src): Bring back isUser, not using flipping

Cloud7050 committed:
FEAT(): Hook real logic back up to UI

Cloud7050 committed:
devfeat(checkstyle): Readd ParameterName but allow leading _

Cloud7050 committed:
Merge branch 'branch-Level-10'

Cloud7050 committed:
chore(): Scrap text-based tests

Cloud7050 committed:
feat(): Use Java assertions

There are some commented assumptions that could be verified using
Java assertions.

Let's turn them into Java assertions, and update the Gradle run task to enable
assertions.

This commit also makes some tweaks to the Gradle config.

Cloud7050 committed:
chore(gradle): Tidy config

Cloud7050 committed:
Merge pull request #2 from Cloud7050/branch-A-Assertions

Add assertions

Cloud7050 committed:
Merge branch 'master' into branch-A-CodeQuality

Cloud7050 committed:
Merge pull request #3 from Cloud7050/branch-A-CodeQuality

Improve code quality

Cloud7050 committed:
devfeat(github): Add action

Cloud7050 committed:
chore(gradle): Update gradlew and readd sh script

Cloud7050 committed:
fix(gradle): Set executable bit of gradlew

Cloud7050 committed:
devfeat(test): Support UTC tests

Cloud7050 committed:
chore(test): Remove unused import

Cloud7050 committed:
fix(test): Fix timestamp for UTC variant

Cloud7050 committed:
docs(): Add screenshot

Cloud7050 committed:
docs(readme): Add image and UG link

Cloud7050 committed:
docs(readme): Add alt link to GitHub Page

Cloud7050 committed:
docs(readme): Reorder paras due to tall image

Cloud7050 committed:
docs(readme): Add real user guide

Cloud7050 committed:
feat(src): Polish messages

Cloud7050 committed:
feat(fxml): Shrink profile pics

Cloud7050 committed:
fix(fxml): Fix messages overflowing X

Cloud7050 committed:
feat(): Change completion text to emoji

Now that output is no longer to the CLI, where emoji don't really work
properly, this feature can be brought back.

Cloud7050 committed:
Revert "feat(): Change completion text to emoji"

This reverts commit e8ad23a7affc620b107756597e6ab411635985f3.

Cloud7050 committed:
docs(): Update showcase image

Cloud7050 committed:
docs(): Header capitalisation
